// @squareboat/nest-console/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const pc = require("picocolors");
const Table = require("cli-table3");
class Logger {
    static info(msg) {
        console.log(pc.cyan(msg));
    }
    static warn(msg) {
        console.log(pc.yellow(msg));
    }
    static error(msg) {
        console.log(pc.bgRed(pc.bold(msg)));
    }
    static line() {
        console.log(pc.gray("-".repeat(process.stdout.columns / 2)));
    }
    static success(msg) {
        console.log(pc.green(msg));
    }
    static table(rows) {
        let columns = [];
        for (const row of rows) {
            columns = columns.concat(Object.keys(row));
        }
        columns = [...new Set(columns)];
        const uniqueCols = [];
        for (const col of columns) {
            uniqueCols.push(pc.cyan(pc.bold(col.charAt(0).toUpperCase() + col.slice(1))));
        }
        const pRows = [];
        rows.forEach((r) => pRows.push(Object.values(r).map((e) => (e && e.toString && e.toString()) || "")));
        const p = new Table({ head: uniqueCols });
        p.push(...pRows);
        console.log(p.toString());
    }
}
exports.Logger = Logger;
