// @squareboat/nest-console/dist/module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var ConsoleModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleModule = void 0;
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const commands_1 = require("./commands");
const explorer_1 = require("./explorer");
let ConsoleModule = ConsoleModule_1 = class ConsoleModule {
    static register() {
        return {
            global: true,
            module: ConsoleModule_1,
            imports: [core_1.DiscoveryModule],
            providers: [commands_1.ListCommands, explorer_1.ConsoleExplorer],
        };
    }
    static registerAsync() {
        return {
            global: true,
            module: ConsoleModule_1,
            imports: [core_1.DiscoveryModule],
            providers: [explorer_1.ConsoleExplorer, commands_1.ListCommands],
        };
    }
};
ConsoleModule = ConsoleModule_1 = __decorate([
    (0, common_1.Global)(),
    (0, common_1.Module)({
        imports: [core_1.DiscoveryModule],
        providers: [explorer_1.ConsoleExplorer, commands_1.ListCommands],
    })
], ConsoleModule);
exports.ConsoleModule = ConsoleModule;
